Spack environment

oparry-ukaea opened the pull request:
The main aims of this PR are to 
1. Make hermes-3 and xhermes easily installable via spack, such that the tests all run out-of-the-box.
2. Facilitate future development of hermes-3 in a spack environment.

Details
- Adds BOUT-spack as a git submodule, which includes packages for hermes-3, xhermes and those of its Python dependencies that don't already have a published package.
- Adds a spack.yaml with hermes-3 as the top-level spec (built using gcc, by default)
- Revises the recommended procedure in the sphinx docs for installing with spack
- Adds a wrapper script to define a few useful commands for managing the environment

Resolves #356

cmacmackin reviewed (commented):
Despite what I said earlier, now I think it would make more sense to create a separate repo for the Spack packages. It would serve as a central repository for all `boutproject` packages. Then each package could include it as a submodule.

Related to this, I would suggest writing a separate package for Bout++, rather than just building it as part of Hermes-3. We could still include Bout++ as a submodule here, but build it as a separate package like we did with Nektar++ for NESO.

We could potentially leave those to future PRs.

mikekryjak commented:
Hi @**Owen Parry** I've finally been able to go through Spack docs and take a look at this. I have some questions about this PR:
- I want to install the master branch, but then change the code for development. What's the easiest way to recompile the code without Spack pulling the master branch again? Is it by using `spack develop` or something else?
- If I install it your way, I'll have a Spack python activate in my Hermes-3 Spack environment, right? Is there an easy way to have the Hermes-3 environment activated but not use the Python from there apart from I guess just activating my existing `venv` after the spack environment is activated? 
- Is there an easy way to pass a branch or commit name to Spack to work from instead of what you defined in `spack.yaml`?

mrhardman commented:
A quick note that googletest is used as a submodule of Hermes-3 (e.g., https://github.com/boutproject/BOUT-dev/tree/90f8506da67cfb2332c6933415524fd2c8ead3d3) and BOUT-dev (e.g., https://github.com/boutproject/BOUT-dev/tree/90f8506da67cfb2332c6933415524fd2c8ead3d3/externalpackages), and we may wish to use a version of googletest from Spack in the future.

mikekryjak commented:
I asked @ZedThree and he says he doesn't think there is a way to compile without googletest, but that it probably shouldn't matter because each executable will only see their own version of googletest (as we discussed in the meeting). Could be something to revisit at a later time.

oparry-ukaea commented:
> Related to this, I would suggest writing a separate package for Bout++, rather than just building it as part of Hermes-3. We could still include Bout++ as a submodule here, but build it as a separate package like we did with Nektar++ for NESO.
> 
> We could potentially leave those to future PRs.

Agree that a separate package for BOUT makes sense - I'd prefer to postpone that to a future PR.

oparry-ukaea commented:
@mrhardman @mikekryjak
I'm writing some instructions for this in sphinx.
My inclination is to replace what Michael wrote, rather than document (and support) both approaches.
Is that ok with you both?

mrhardman commented:
> @mrhardman @mikekryjak I'm writing some instructions for this in sphinx. My inclination is to replace what Michael wrote, rather than document (and support) both approaches. Is that ok with you both?

Yes, I am happy with this. If you think it appropriate, in the new documentation that you write you can simply point out that Spack does not preclude a manual install of PETSc, SUNDIALS, BOUT++, etc. Once you have the Spack environment you can also use the lower-level modules (e.g., MPI, netCDF, etc) from that environment to follow the original CMake install instructions if desired.

oparry-ukaea commented:
Hi @bendudson, we've changed our minds on this a few times, so the description of this PR is out of date - I'll fix that now.

The plan we've landed on is for this PR to pull in `BOUT-spack` as a git submodule and to add a `spack.yml` to configure the spack environment.  hermes-3 is treated as a 'develop' package, meaning that it's built using whatever code is currently checked out, rather than letting spack checkout a particular version.

oparry-ukaea commented:
@mrhardman Many thanks for the testing and feedback.
I've changed the installation instructions as suggested and modified the wrapper script to load the 'view' when you activate, which should mean that the right Python interpreter gets picked up when running the tests.
You still need to `spack load cmake` to make ctest available, but other than that, things *should* just work.

cmacmackin commented:
>You still need to spack load cmake to make ctest available, but other than that, things should just work.

If you want that to happen automatically, we could add it to the specs in the environment YAML file.

mrhardman commented:
I have tested the install instructions on 
```
commit ecff39ae5f1ae55abe16bb5253bb067b945ac3b0 (HEAD -> spack-env, origin/spack-env)
```
and found that everything works, and `ctest` was found without loading any modules explicitly.

oparry-ukaea commented:
Thanks Michael, my last two commits don't change the install process or docs, so probably no need to re-test.
Happy to make further changes if others have feedback, otherwise this is ready to merge.

mrhardman commented:
> Thanks Michael, my last two commits don't change the install process or docs, so probably no need to re-test. Happy to make further changes if others have feedback, otherwise this is ready to merge.

I did check again, and the build proceeds as before, the tests pass.

cmacmackin commented:
>would result in ./my_build/hermes-3 being linked to the spack-installed version of libhermes-3.so (which includes none of the changes).

I'm surprised by that. I thought the build-env command created a truly isolated environment that wouldn't be aware of the active view.

cmacmackin commented:
>On closer inspection, you're probably right; the dependencies-only view doesn't help, unfortunately.
The first two entries in the rpath of the exec(s) always seem to be `$(spack location -i hermes-3)/lib:$(spack location -i hermes-3)/lib64`, regardless of whether hermes-3 is in the view.

On my machine, the dynamic linker still picks up the adjacent libhermes-3-lib.so before it looks at the RPATH. E.g.:
```
$ ldd build/hermes-3 
	linux-vdso.so.1 (0x000073c66e2ca000)
	libhermes-3-lib.so => /home/cmacmack/Documents/hermes-3/build/libhermes-3-lib.so (0x000073c66de00000)
	libbout++.so.5.1.0 => /home/cmacmack/Documents/hermes-3/build/external/BOUT-dev/lib/libbout++.so.5.1.0 (0x000073c66d600000)
	libpetsc.so.3.22 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/petsc-3.22.1-okmxebgq57neg42mcenwfppckt2b63td/lib/libpetsc.so.3.22 (0x000073c66c000000)
	libfmt.so.10 => /home/cmacmack/Documents/hermes-3/build/external/BOUT-dev/lib/libfmt.so.10 (0x000073c66e1a6000)
	libmpi.so.12 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/mpich-4.2.3-3fk7c23c3vxizv32u4h35bkgsm5ne62m/lib/libmpi.so.12 (0x000073c66b600000)
	libstdc++.so.6 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/gcc-runtime-13.2.0-tpxmqehkicdw2ty3z7flkdvy2kn54v6e/lib/libstdc++.so.6 (0x000073c66b200000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000073c66e099000)
	libgcc_s.so.1 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/gcc-runtime-13.2.0-tpxmqehkicdw2ty3z7flkdvy2kn54v6e/lib/libgcc_s.so.1 (0x000073c66e06b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000073c66ae00000)
	libpvpre.so.1.0.0 => /home/cmacmack/Documents/hermes-3/build/external/BOUT-dev/lib/libpvpre.so.1.0.0 (0x000073c66e062000)
	libnetcdf-cxx4.so.1 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/netcdf-cxx4-4.3.1-palqp5b55vg7foq2xcfsacsaqejnnpgb/lib/libnetcdf-cxx4.so.1 (0x000073c66dda7000)
	libfftw3.so.3 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/fftw-3.3.10-fgqn23tifj7pbl4gdfky77rrnu3ie4pf/lib/libfftw3.so.3 (0x000073c66aa00000)
	libpvode.so.1.0.0 => /home/cmacmack/Documents/hermes-3/build/external/BOUT-dev/lib/libpvode.so.1.0.0 (0x000073c66dd95000)
	libsundials_nvecparallel.so.7 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/sundials-7.1.1-dt37q2vxmhhlghzgmqtdylbq6ilygtik/lib/libsundials_nvecparallel.so.7 (0x000073c66e053000)
	libsundials_cvode.so.7 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/sundials-7.1.1-dt37q2vxmhhlghzgmqtdylbq6ilygtik/lib/libsundials_cvode.so.7 (0x000073c66dd5a000)
	libsundials_ida.so.7 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/sundials-7.1.1-dt37q2vxmhhlghzgmqtdylbq6ilygtik/lib/libsundials_ida.so.7 (0x000073c66dd25000)
	libsundials_arkode.so.6 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/sundials-7.1.1-dt37q2vxmhhlghzgmqtdylbq6ilygtik/lib/libsundials_arkode.so.6 (0x000073c66b559000)
	libsundials_core.so.7 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/sundials-7.1.1-dt37q2vxmhhlghzgmqtdylbq6ilygtik/lib/libsundials_core.so.7 (0x000073c66dd0a000)
	libHYPRE-2.32.0.so => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/hypre-2.32.0-ehouvbckzzwp3vpcfsvw5d4cvpt4reax/lib/libHYPRE-2.32.0.so (0x000073c66a600000)
	libopenblas.so.0 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/openblas-0.3.28-nmku5ummj2765su2mejzvpidm6cukeuv/lib/libopenblas.so.0 (0x000073c668db8000)
	libparmetis.so => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/parmetis-4.0.3-tcqpqml555wi65w6ju7whjnwve22bjxa/lib/libparmetis.so (0x000073c66dcbf000)
	libmetis.so => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/metis-5.1.0-giloac6yeov6vw62i5p5mg3vo7pfsjni/lib/libmetis.so (0x000073c66bf8b000)
	libhdf5.so.310 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/hdf5-1.14.5-5wnezwcrkpsvh5ija3qqot4woem24gpb/lib/libhdf5.so.310 (0x000073c668800000)
	/lib64/ld-linux-x86-64.so.2 (0x000073c66e2cc000)
	libhwloc.so.15 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/hwloc-2.11.1-e5kzon76u5cpvvwlufsohooxnyx4iztt/lib/libhwloc.so.15 (0x000073c66b4f4000)
	libyaksa.so.0 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/yaksa-0.3-tll5us37wjjq6ar5nsqdmmx4br5dd4uh/lib/libyaksa.so.0 (0x000073c666400000)
	libfabric.so.1 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/libfabric-1.22.0-73gymhxw5s6doei3xyftd3kgkivjnbrf/lib/libfabric.so.1 (0x000073c66b14d000)
	libnetcdf.so.19 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/netcdf-c-4.9.2-kz2wo2nxuvblkpstczroqbjtx5hstxgw/lib/libnetcdf.so.19 (0x000073c66621a000)
	libgfortran.so.5 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/gcc-runtime-13.2.0-tpxmqehkicdw2ty3z7flkdvy2kn54v6e/lib/libgfortran.so.5 (0x000073c665e00000)
	libz.so.1 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/zlib-ng-2.2.1-pnbfqqo7z5is3owtnkd5gl3cqd3v4cvu/lib/libz.so.1 (0x000073c66b4cd000)
	libpciaccess.so.0 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/libpciaccess-0.17-mrbyka33v42wndnplxzdpsncmymrjke6/lib/libpciaccess.so.0 (0x000073c66dcaf000)
	libxml2.so.2 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/libxml2-2.13.4-pif3uvweiyorqxtpvbuqelcn53rowy7a/lib/libxml2.so.2 (0x000073c665ca8000)
	libhdf5_hl.so.310 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/hdf5-1.14.5-5wnezwcrkpsvh5ija3qqot4woem24gpb/lib/libhdf5_hl.so.310 (0x000073c66b4a6000)
	liblzma.so.5 => /opt/spack/opt/spack/linux-linuxmint22-skylake/gcc-13.2.0/xz-5.4.6-qxaedv53wcqtjchnvghuxaw3avmxp7mb/lib/liblzma.so.5 (0x000073c66b11f000)

```

oparry-ukaea commented:
Thank Chris, much appreciated.
I've reverted the view change and added that workaround for hermes-3-lib, hermes-3 and hermes_unit_tests.

oparry-ukaea commented:
Tests seem to be failing on CI because the [latest release of xarray is broken](https://github.com/pydata/xarray/issues/10412). Looks like it should be fixed imminently, though.

oparry-ukaea commented:
New version of xarray released, so the tests pass again.

cmacmackin commented:
@oparry-ukaea I've created a boutproject/BOUT-spack#5 which you may want to merge and use on this branch.

oparry-ukaea commented:
Thanks @cmacmackin, nice work tracking down the libintl/gettext issue, I'll take a look.

oparry-ukaea commented:
This is ready to merge now.
